hail/src/main/scala/is/hail/expr/ir/ForwardLets.scala

```diff
@@ -0,0 +1,77 @@
+package is.hail.expr.ir
+
+case class RefCounter(var maxNestingLevel: Int = 0, var nRef: Int = 0) {
+  def register(level: Int) {
+    maxNestingLevel = math.max(level, maxNestingLevel)
+    nRef += 1
+  }
+}
+
+object ForwardLets {
+  def analyzeRefsByNestingLevel(ir: BaseIR, binding: String, nestingLevel: Int, counter: RefCounter): Unit = {
```

Comment:
I kinda think this should be called `evaluation_count` and it should return a `Option[Int]`. Abstractly, this is estimating the number of times a bound name is evaluated. `None` means "0 or more times". `Some(i)` means "exactly i times".

I think this framing is both useful in general outside of this particular optimization pass and is also more robust. If new IR nodes are added it's much clearer how this method needs to change. (Aside: for this reason, I also prefer explicitly listing all the nodes in the match and requiring no new match failures (we could enforce this by logging a grep of the compileScala output and doing a diff on each PR build)).

Reply:
this computes two metrics, though - both the number of uses and the maximum nesting level.

Reply:
I can push the [Bindings code](https://github.com/hail-is/hail/pull/5041/files#diff-60a91e9c881938634a3a79265e59bc89R32) from #5041 into this PR and assert that, in the base case, the IR binds no variables.

Would that be sufficient?

Reply:
oh, do you mean that anything with binding level > 0 would return `None`?